doc/tut1.rst

```diff
@@ -1701,15 +1722,17 @@ Example:
 
 .. code-block:: nim
     :test: "nim c $1"
-  proc echoItem(x: int) = echo x
+  proc greet(name: string): string =
+    result = "Hello, " & name & "!"
+
+  proc bye(name: string): string =
+    result = "Goodbye, " & name & "."
```

Comment:
Something that has bugged me for a while: is there any particular reason for the `return =`? One could just directly return the expression.

Reply:
for secure code, we strongly recommend either `return` or expressions - the usage or `result` has been a source of numerous bugs that could have been avoided with the stronger guarantees that expressions / explicit return give: https://status-im.github.io/nim-style-guide/language.result.html

Reply:
You keep blaming `result` for the lack of definite assignment enforcement.

Reply:
even DAA doesn't protect against multiple overwriting assignment, which is one category of bugs we've found - expression return prevents this as well as preventing "forgotten" return paths.

Also, as long as DAA doesn't exist, we strive to create high-quality code using the tools available - "use this feature because maybe in the future, it'll work" doesn't cut it in production use - what happens though is that more buggy nim code is put out there in the meantime.

Reply:
besides, typing `result =` in these examples is just noise, like @konsumlamm points out (typo aside) - it doesn't really help the code in any way.

Reply:
DAA does exist and I've patched large parts of the stdlib (and the compiler itself) once to enforce it. Putting more work into DAA would avoid the "result used warning" **language dialect**.

Reply:
`dialect` is generally used for features that are mutually incompatible - for example, you can't use a library that doesn't support DAA (because DAA enforces rules that don't exist in the current version of the language) in an application that requires it for QA. The best you'll be able to do with DAA in the real world is a warning, I suspect, because there's plenty of buggy code out there that violates it, but there's also _some_ code that violates it and is not buggy - sometimes code relies on the default zero-init.

`result` or not is a purely stylistic choice - you choose to use a more constrained construct (expressions) to gain the peace of mind that you won't shoot yourself in the foot - it has no impact whatsoever on compatibility with other code, ergo it is not a dialect - all code written without `result` can be used in an application that uses `result`. Crucially, if you use expression returns exclusively, _you can't trigger DAA_ - your code is already compatible with both DAA and not modes.

To this point, if std lib had been written with expression returns, you'd have less to fix and fewer bugs in it when introducing DAA.

Reply:
Well I cannot use `result` if the warning is enabled so it's pretty much a dialect. More importantly, `result` is the idiomatic way to get an accumulator and neither `return` nor expression based programming can give you the same. So for an accumulator you would use `var result: T; loop: result.add x; return result` which is much more error prone than using the "real" result.

Reply:
> if the warning is enabled

our hope is that we can enable the warning per library, so that _that_ library is free of usages - it's a bit like we've done with exceptions: we isolate them such that they don't leak outside of well-defined boundaries.